### PR TITLE
refactor(cli): Introduce v1.Interface for CLI. Closes #2107

### LINKED ISSUE #2107: Error while using rc6
Warning: You tried to redirect to the same route you're currently on: "workflows"
(anonymous) @ main.9263c6e00044a0788ca2.js:12
r @ main.9263c6e00044a0788ca2.js:12
t.componentDidUpdate @ main.9263c6e00044a0788ca2.js:17
Im @ main.9263c6e00044a0788ca2.js:111
P_ @ main.9263c6e00044a0788ca2.js:111
e @ main.9263c6e00044a0788ca2.js:98
c @ main.9263c6e00044a0788ca2.js:98
y @ main.9263c6e00044a0788ca2.js:98
F_ @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:111
t.unstable_runWithPriority @ main.9263c6e00044a0788ca2.js:111
Yy @ main.9263c6e00044a0788ca2.js:111
Dy @ main.9263c6e00044a0788ca2.js:111
Sy @ main.9263c6e00044a0788ca2.js:111
Ly @ main.9263c6e00044a0788ca2.js:111
wy @ main.9263c6e00044a0788ca2.js:111
K_ @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:111
Fy @ main.9263c6e00044a0788ca2.js:111
Ny @ main.9263c6e00044a0788ca2.js:111
Ky.render @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:111
Cy @ main.9263c6e00044a0788ca2.js:111
tb @ main.9263c6e00044a0788ca2.js:111
render @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:82
(anonymous) @ main.9263c6e00044a0788ca2.js:82
n @ main.9263c6e00044a0788ca2.js:1
(anonymous) @ main.9263c6e00044a0788ca2.js:1
(anonymous) @ main.9263c6e00044a0788ca2.js:1

Warning: You tried to redirect to the same route you're currently on: "workflows"
(anonymous) @ main.9263c6e00044a0788ca2.js:12
r @ main.9263c6e00044a0788ca2.js:12
t.componentDidUpdate @ main.9263c6e00044a0788ca2.js:17
Im @ main.9263c6e00044a0788ca2.js:111
P_ @ main.9263c6e00044a0788ca2.js:111
e @ main.9263c6e00044a0788ca2.js:98
c @ main.9263c6e00044a0788ca2.js:98
y @ main.9263c6e00044a0788ca2.js:98
F_ @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:111
t.unstable_runWithPriority @ main.9263c6e00044a0788ca2.js:111
Yy @ main.9263c6e00044a0788ca2.js:111
Dy @ main.9263c6e00044a0788ca2.js:111
Sy @ main.9263c6e00044a0788ca2.js:111
Ly @ main.9263c6e00044a0788ca2.js:111
wy @ main.9263c6e00044a0788ca2.js:111
K_ @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:111
Fy @ main.9263c6e00044a0788ca2.js:111
Ny @ main.9263c6e00044a0788ca2.js:111
Ky.render @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:111
Cy @ main.9263c6e00044a0788ca2.js:111
tb @ main.9263c6e00044a0788ca2.js:111
render @ main.9263c6e00044a0788ca2.js:111
(anonymous) @ main.9263c6e00044a0788ca2.js:82
(anonymous) @ main.9263c6e00044a0788ca2.js:82
n @ main.9263c6e00044a0788ca2.js:1
(anonymous) @ main.9263c6e00044a0788ca2.js:1
(anonymous) @ main.9263c6e00044a0788ca2.js:1

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the README.
* [x] I've signed the CLA and required builds are green. 

See #1951
Closes #2107

Anyone reviewing this - we should sit down so we can talk over the solution.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -381,9 +381,10 @@ test-api: test-images
 	go test -timeout 3m -v -count 1 -p 1 -run ArgoServerSuite ./test/e2e
 
 .PHONY: test-cli
-test-cli: test-images
+test-cli: test-images cli
 	# Run CLI tests
 	go test -timeout 1m -v -count 1 -p 1 -run CliSuite ./test/e2e
+	go test -timeout 1m -v -count 1 -p 1 -run CLIWithServerSuite ./test/e2e
```
Comment (simster7) [inline Makefile]: Very minor but it should be `CliWithServerSuite`

File: cmd/argo/commands/common.go [hunk 2]
```diff
@@ -14,27 +14,31 @@ import (
...
 	"github.com/argoproj/argo/workflow/templateresolution"
 )
 
 // Global variables
 var (
-	restConfig       *rest.Config
-	clientset        *kubernetes.Clientset
-	wfClientset      *versioned.Clientset
-	wfClient         v1alpha1.WorkflowInterface
+	// DEPRECATED
```
Comment (simster7) [inline cmd/argo/commands/common.go]: Why can't these just be deleted? AFAIK, people don't import our CLI

File: docs/argo-server.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# Argo Server
+
+![alpha](assets/alpha.svg)
+
+> v2.5 and after
+
+The Argo Server is a server that exposes an API and UI for workflows. You'll need to use this if you want to use [offload large workflows](offloading-large-workflows.md) or the [workflow archive](workflow-archive.md).
```
Comment (simster7) [inline docs/argo-server.md]: Minor grammar
```suggestion
The Argo Server is a server that exposes an API and UI for workflows. You'll need to use this if you want to [offload large workflows](offloading-large-workflows.md) or the [workflow archive](workflow-archive.md).
```

File: docs/argo-server.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+It replaces the Argo UI.
+
+## Hosted Mode
+
+Use this mode if:
+
+* You want a drop-in replacement for the Argo UI.
+* If you need to prevent users from directly accessing the database.
+
+Hosted mode is provide  as part of the standard [manifests](../manifests), [specifically in argo-server-deployment.yaml](../manifests/base/argo-server/argo-server-deployment.yaml) .
```
Comment (simster7) [inline docs/argo-server.md]:
```suggestion
Hosted mode is provided as part of the standard [manifests](../manifests), [specifically in argo-server-deployment.yaml](../manifests/base/argo-server/argo-server-deployment.yaml) .
```

File: docs/cli.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# CLI
+
+Most help topics are provide using built-in help
```
Comment (simster7) [inline docs/cli.md]:
```suggestion
Most help topics are provided using built-in help
```

File: pkg/apiclient/apiclient.go [hunk 6]
```diff
@@ -1,6 +1,23 @@
...
+	NewArchivedWorkflowServiceClient() (workflowarchivepkg.ArchivedWorkflowServiceClient, error)
+	NewWorkflowServiceClient() workflowpkg.WorkflowServiceClient
+}
+
+func NewClient(argoServer, token string, clientConfig clientcmd.ClientConfig) (context.Context, Client, error) {
+	if argoServer != "" {
+		return newArgoServerClient(argoServer, token)
+	} else {
+		return newClassicClient(clientConfig)
+	}
```
Comment (simster7) [inline pkg/apiclient/apiclient.go]: 🎊

File: test/e2e/fixtures/e2e_suite.go [hunk 7]
```diff
@@ -251,9 +244,9 @@ func (s *E2ESuite) printPodLogs(logCtx *log.Entry, namespace, pod, container str
 	fmt.Println("---")
 }
 
-func (s *E2ESuite) Given() *Given {
+func (s *E2ESuite) Given(t *testing.T) *Given {
```
Comment (simster7) [inline test/e2e/fixtures/e2e_suite.go]: Why is this necessary?

Comment (alexec) [reply test/e2e/fixtures/e2e_suite.go]: to correctly report errors to sub-tests

File: cmd/argo/commands/resubmit.go [hunk 8]
```diff
@@ -40,7 +40,7 @@ func NewResubmitCommand() *cobra.Command {
 				defer conn.Close()
 				apiGRPCClient, ctx := GetWFApiServerGRPCClient(conn)
 				errors.CheckError(err)
-				wfReq := workflow.WorkflowGetRequest{
+				wfReq := workflowpkg.WorkflowGetRequest{
 					Namespace: namespace,
 					Name:      args[0],
 				}
```
Comment (sarabala1979) [inline cmd/argo/commands/resubmit.go]: The rest of the implementation still remains the old way of doing.  Can it not be refactored?

File: cmd/argo/commands/resume.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ func NewResumeCommand() *cobra.Command {
 				conn := client.GetClientConn()
```
Comment (sarabala1979) [inline cmd/argo/commands/resume.go]: Same as previous comment

File: cmd/argo/commands/watch.go [hunk 10]
```diff
@@ -38,7 +38,7 @@ func apiServerWatchWorkflow(wfName string) {
 	defer conn.Close()
 	apiClient, ctx := GetWFApiServerGRPCClient(conn)
 	fieldSelector := fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", wfName))
-	wfReq := workflow.WatchWorkflowsRequest{
+	wfReq := workflowpkg.WatchWorkflowsRequest{
 		Namespace: namespace,
 		ListOptions: &metav1.ListOptions{
 			FieldSelector: fieldSelector.String(),
```
Comment (sarabala1979) [inline cmd/argo/commands/watch.go]: Same as the previous comment

File: cmd/argo/commands/resubmit.go (hunk 8 above)
Comment (alexec) [reply cmd/argo/commands/resubmit.go]: I'd like to do it piecemeal - as a lot of the CLI code has no automated test - so refactoring is more likely to introduce bugs and regressions. 

Whenever we need to change the CLI, we should complete the appropriate part of the client and migrate.

File: cmd/argo/commands/common.go (hunk 2 above)
Comment (alexec) [reply cmd/argo/commands/common.go]: our CLI uses these methods - this is really more to make it easy for us to identify methods to remove
